src/torrent.cpp

```diff
@@ -5664,99 +5675,66 @@ namespace {
 	void torrent::on_file_priority(disk_io_job const* j)
 	{
 		dec_refcount("file_priority");
+
 		boost::scoped_ptr<std::vector<boost::uint8_t> > p(j->buffer.priorities);
-		if (m_file_priority == *p) return;
 
-		// in this case, some file priorities failed to get set
-		m_file_priority = *p;
-		update_piece_priorities();
+		if (j->ret == piece_manager::no_error)
+		{
+			TORRENT_ASSERT(!j->error);
+			m_file_priority = *p;
```

Comment:
do we even need to assign the priorities in this case?

Reply:
This change actually fixes the issue.

Reply:
I think we need to assign this unconditionally here. (i.e. hoist it out of the if-block). Even if there is an error reported from the disk thread, some priorities may have been applied.

Reply:
I'm not quite sure that this must be done, since `file_error_alert` and `torrent_error_alert` are not __critical__. That mean they could be lost and user won't have a chance to handle priority deviation.

Reply:
I found a problem with this, `disk_io_thread::do_file_priority` always returns `no_error`, thus second branch will never be executed.

Reply:
oh, I think it was intended to propagate file errors.
Whether the user receives the alert or not is somewhat separate. I think it's important for libtorrent's internal state to stay consistent regardless.
The user can always ask for the priorities.